Support matching Args for SubgraphMatcher

SherlockNoMad opened the pull request:
Subgraph matcher now handles the matching of non-Node arguments. 

Here are the 4 cases
- pn is Node, gn is Node: this go through the regular _match_node() function
- pn is Noed, gn is not a Node: this is a match if only pn is a placeholder op
- pn is not Node, gn is Node: this is a no match case
- pn is not a Node, gn is not a Node: this will go through the argument comparison. 

With this change
```
def target(x):
    return foo(x, 3)

def pattern(x, y):
    return foo(x, y) 
```

is a match

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #85456

SherlockNoMad committed:
Support matching Args for SubgraphMatcher

[ghstack-poisoned]

SherlockNoMad committed:
Update on "Support matching Args for SubgraphMatcher"




[ghstack-poisoned]

SherlockNoMad commented:
@pytorchbot merge -g

jerryzh168 commented on torch/fx/passes/utils/matcher_utils.py:
```diff
@@ -146,9 +167,23 @@ def _match_nodes(self, pn: Node, gn: Node, match: InternalMatch) -> bool:
+        if len(pn_flatten_args) == len(gn_flatten_args):
+            for pn_, gn_ in zip(pn_flatten_args, gn_flatten_args):
+                if isinstance(gn_, Node) and isinstance(pn_, Node):
+                    matched = self._match_nodes(pn_, gn_, match)
```
nit: maybe we can just use `match_found` here

SherlockNoMad replied:
In this context, matched and match_found carries different meaning... 

`matched` is for each argument, `match_found` is for current node
if and only if `matched` for all the argument is True, then `match_found` is True.

jerryzh168 commented on torch/fx/passes/utils/matcher_utils.py:
```diff
@@ -124,7 +125,27 @@ def _remove_overlapping_matches(self, matches: List[InternalMatch]) -> List[Inte
+        elif not isinstance(pn, Node) and isinstance(gn, Node):
+            return False
+        else:
+            return type(gn) == type(pn) and gn == pn
```
does `gn == pn` imply `type(gn) == type(pn)`?

jerryzh168 replied:
oh looks like not

jerryzh168 reviewed (approved):
thanks for the fix!

jerryzh168 commented:
also might be better to add a summary message

SherlockNoMad commented:
> also might be better to add a summary message

Thanks for reviewing this! 
I have updated summary message, merging now...

SherlockNoMad commented:
@pytorchbot merge -g

SherlockNoMad committed:
Update on "Support matching Args for SubgraphMatcher"

Subgraph matcher now handles the matching of non-Node arguments. 

Here are the 4 cases
- pn is Node, gn is Node: this go through the regular _match_node() function
- pn is Noed, gn is not a Node: this is a match if only pn is a placeholder op
- pn is not Node, gn is Node: this is a no match case
- pn is not a Node, gn is not a Node: this will go through the argument comparison. 

With this change
```
def target(x):
    return foo(x, 3)

def pattern(x, y):
    return foo(x, y) 
```

is a match




[ghstack-poisoned]

SherlockNoMad commented:
@pytorchbot merge -g
